clients/src/main/java/org/apache/kafka/common/utils/Utils.java

```diff
@@ -781,4 +781,29 @@ public static long computeChecksum(ByteBuffer buffer, int start, int size) {
         return Crc32.crc32(buffer.array(), buffer.arrayOffset() + start, size);
     }
 
+    /**
+     * Read data from the channel to the given byte buffer until there are no bytes remaining in the buffer or
+     * the channel has reached end of stream
+     * @param channel File channel containing the data to read from
+     * @param destinationBuffer The buffer into which bytes are to be transferred
+     * @param position The file position at which the transfer is to begin; must be non-negative
+     *
+     * @throws IllegalArgumentException If position is negative
+     * @throws IOException If reaching EOF or an I/O error occurs. See {@link FileChannel#read(ByteBuffer, long)}
+     */
+    public static void readFully(FileChannel channel, ByteBuffer destinationBuffer, long position) throws IOException {
+        if (position < 0) {
+            throw new IllegalArgumentException("The file channel position cannot be negative, but we got " + position);
+        }
+        long currentPosition = position;
+        int bytesRead;
+        do {
+            bytesRead = channel.read(destinationBuffer, currentPosition);
+            currentPosition += bytesRead;
+        } while (bytesRead != -1 && destinationBuffer.hasRemaining());
+        if (bytesRead == -1) {
+            throw new IOException(String.format("Reached EOF before filling up the buffer. Current file channel position = %d, buffer remaining = %d",
```

Comment:
I am not sure what is your concrete suggestion. Our aim here is to provide internal utility methods that do the right thing concisely and with good error messages.

I agree that it may make sense to have two methods. One could be `readFullyOrFail` and the other could simply be `readFully` maybe. The former would throw the exception on EOF while the other would not.

To clarify my suggestion, I meant that the `readFullyOrFail` variant could take a string parameter that would be included in the IOException message. It's useful to know that we failed to read "magic string", for example.